toolchain/check/context.cpp

```diff
@@ -395,18 +395,18 @@ static auto GetCorePackage(Context& context, SemIRLoc loc)
   auto core_inst_id = context.LookupNameInExactScope(
       loc, core_name_id, SemIR::NameScopeId::Package,
       context.name_scopes().Get(SemIR::NameScopeId::Package));
-  if (!core_inst_id.is_valid()) {
-    context.DiagnoseNameNotFound(loc, core_name_id);
-    return SemIR::NameScopeId::Invalid;
+  if (core_inst_id.is_valid()) {
+    // We expect it to be a namespace.
+    if (auto namespace_inst =
+            context.insts().TryGetAs<SemIR::Namespace>(core_inst_id)) {
+      return namespace_inst->name_scope_id;
+    }
   }
 
-  // We expect it to be a namespace.
-  if (auto namespace_inst =
-          context.insts().TryGetAs<SemIR::Namespace>(core_inst_id)) {
-    return namespace_inst->name_scope_id;
-  }
-  // TODO: This should really diagnose the name issue.
-  context.DiagnoseNameNotFound(loc, core_name_id);
+  CARBON_DIAGNOSTIC(CoreNotFound, Error,
+                    "Package `Core` implicitly referenced here but not found.",
+                    SemIR::NameId);
```

Comment:
My thinking was: if lookup for `Core` finds something that's not even a namespace, then it definitely didn't find a package named `Core`, so the same diagnostic would still make sense.

Reply:
Should it proactively identify that it's looking at package then, and not just an arbitrary namespace named `Core`? You could do that by ensuring the namespace has import IRs.

Reply:
I'm on the fence. If someone has turned off the implicit prelude import and defined a local namespace named `Core`, it seems fine to me for that to work. But maybe that's not something we want to support, I'm not sure. Either way this seems like something it'd be good to have a test for, if we don't already.

But... this behavior is not being changed in this PR. I'm only adding a more precise diagnostic for the existing failures in this PR, not changing which cases get diagnosed.

Reply:
A TODO is being removed, maybe it should be kept in some form? The thing is, a name `Core` _can_ be found, and the diagnostic may be confusing if:

- Core is not a SemIR::Namespace.
- Core is a namespace but not a package.

For me, this change is peculiar because the main thrust of the change is `As`, but it also changes how `Core` name lookup is diagnosed. Maybe that should also be noted in the PR commit description, as I think about that?

Reply:
Added the TODO back and moved out to #4276.